packages/ses/src/module-load.js

```diff
@@ -176,37 +190,94 @@ const loadWithoutErrorAnnotation = async (
     compartment,
     moduleSpecifier,
     staticModuleRecord,
+    queue,
+    loading,
   );
 };
 
-// `load` asynchronously loads `StaticModuleRecords` and creates a complete
-// graph of `ModuleCompartmentRecords`.
-// The module records refer to each other by a reference to the dependency's
-// compartment and the specifier of the module within its own compartment.
-// This graph is then ready to be synchronously linked and executed.
-export const load = async (
+const memoizedLoadWithErrorAnnotation = async (
   compartmentPrivateFields,
   moduleAliases,
   compartment,
   moduleSpecifier,
+  queue,
+  loading,
 ) => {
-  return promiseCatch(
+  const { name: compartmentName } = weakmapGet(
+    compartmentPrivateFields,
+    compartment,
+  );
+
+  // Prevent data-lock from recursion into branches visited in dependent loads.
+  let compartmentLoading = mapGet(loading, compartment);
+  if (compartmentLoading === undefined) {
+    compartmentLoading = new Map();
+    mapSet(loading, compartment, compartmentLoading);
+  }
+  let moduleLoading = mapGet(compartmentLoading, moduleSpecifier);
+  if (moduleLoading !== undefined) {
+    return moduleLoading;
+  }
+
+  moduleLoading = promiseCatch(
     loadWithoutErrorAnnotation(
       compartmentPrivateFields,
       moduleAliases,
       compartment,
       moduleSpecifier,
+      queue,
+      loading,
     ),
     error => {
-      const { name } = weakmapGet(compartmentPrivateFields, compartment);
       // eslint-disable-next-line @endo/no-polymorphic-call
       assert.note(
         error,
         d`${error.message}, loading ${q(moduleSpecifier)} in compartment ${q(
-          name,
+          compartmentName,
         )}`,
       );
       throw error;
     },
   );
+
+  mapSet(compartmentLoading, moduleSpecifier, moduleLoading);
+
+  return moduleLoading;
+};
+
+/*
+ * `load` asynchronously gathers the `StaticModuleRecord`s for a module and its
+ * transitive dependencies.
+ * The module records refer to each other by a reference to the dependency's
+ * compartment and the specifier of the module within its own compartment.
+ * This graph is then ready to be synchronously linked and executed.
+ */
+export const load = async (
+  compartmentPrivateFields,
+  moduleAliases,
+  compartment,
+  moduleSpecifier,
+) => {
+  /** @type {Set<Promise<Record>>} */
+  const queue = new Set();
+  /** @type {Map<Object, Map<string, Promise<Record>>} */
+  const loading = new Map();
+
+  setAdd(
+    queue,
+    memoizedLoadWithErrorAnnotation(
+      compartmentPrivateFields,
+      moduleAliases,
+      compartment,
+      moduleSpecifier,
+      queue,
+      loading,
+    ),
+  );
+
+  // Drain queue
+  for (const job of queue) {
+    // eslint-disable-next-line no-await-in-loop
+    await job;
+  }
```

Comment:
This is where the await Promise.all went.

Reply:
no more ~~parallelism~~ concurrency?

Reply:
This routine continues to branch concurrency. This is the just final await for every branch.